File: Day 8/demo_async.py
"""
Python Asynchronous Programming Demo

To run this file:
1. Open your terminal in the "Day 8" directory.
2. Type: python demo_async.py
"""

import asyncio
import time

# Defining an async function (Coroutine) in Python using `async def`.
# This function simulates an API query that takes 1.5 seconds.
async def fetch_from_api(resource_name, delay_seconds):
    # asyncio.sleep is a non-blocking wait. It suspends this function and
    # returns control to the event loop so other tasks can run in the meantime.
    await asyncio.sleep(delay_seconds)
    return {"data": f"Result of {resource_name}"}

async def main():
    print("==================================================")
    print("Python asyncio Concurrency Demonstration")
    print("==================================================")

    # --- SCENARIO 1: SEQUENTIAL EXECUTION ---
    print("\n[1] Running Tasks SEQUENTIALLY:")
    start_time = time.time()

    # We await the first API task to complete before starting the second.
    # Total expected delay: 1.5s + 1.5s = ~3.0 seconds total.
    user = await fetch_from_api("User Profile", 1.5)
    print("  - Successfully fetched:", user["data"])

    orders = await fetch_from_api("User Orders", 1.5)
    print("  - Successfully fetched:", orders["data"])

    sequential_duration = time.time() - start_time
    print(f"Sequential execution took: {sequential_duration:.2f} seconds\n")


    # --- SCENARIO 2: CONCURRENT EXECUTION ---
    print("[2] Running Tasks CONCURRENTLY (Parallel):")
    start_parallel_time = time.time()

    # In Python, we trigger concurrency by packaging the coroutines.
    # We do NOT await them separately. We run them concurrently using `asyncio.gather()`.
    # Both start running in the background event loop at the exact same time!
    task1 = fetch_from_api("User Profile (Parallel)", 1.5)
    task2 = fetch_from_api("User Orders (Parallel)", 1.5)

    # asyncio.gather executes both tasks concurrently and waits for both to complete.
    # Total expected time: Only ~1.5 seconds total!
    res1, res2 = await asyncio.gather(task1, task2)
    print("  - Successfully fetched (Parallel):", res1["data"])
    print("  - Successfully fetched (Parallel):", res2["data"])

    parallel_duration = time.time() - start_parallel_time
    print(f"Parallel execution took: {parallel_duration:.2f} seconds")
    print("==================================================\n")

# In Python, async functions cannot be run directly like standard functions.
# We must start the asyncio event loop and run our entrypoint `main` function through it.
if __name__ == "__main__":
    asyncio.run(main())

File: Day 8/demo_async.js
/**
 * Node.js Asynchronous Programming Demo
 * 
 * To run this file:
 * 1. Open your terminal in the "Day 8" directory.
 * 2. Type: node demo_async.js
 */

// A simple helper function that simulates a database or API query.
// It returns a promise that resolves with the data after a specified time.
const fetchFromAPI = (resourceName, delayMs) => {
  return new Promise((resolve) => {
    // setTimeout is asynchronous. Node will run other code while this timer ticks.
    setTimeout(() => {
      resolve({ data: `Result of ${resourceName}` });
    }, delayMs);
  });
};

async function main() {
  console.log("==================================================");
  console.log("🤖 Node.js Async/Await Concurrency Demonstration");
  console.log("==================================================");

  // --- SCENARIO 1: SEQUENTIAL EXECUTION (Blocking Style) ---
  console.log("\n🐢 [1] Running Tasks SEQUENTIALLY:");
  const startTime = Date.now();

  // We await the first API call to finish before we even START the second call.
  // Total expected delay: 1.5 seconds + 1.5 seconds = ~3 seconds total.
  const user = await fetchFromAPI("User Profile", 1500);
  console.log("  - Successfully fetched:", user.data);

  const orders = await fetchFromAPI("User Orders", 1500);
  console.log("  - Successfully fetched:", orders.data);

  const sequentialDuration = (Date.now() - startTime) / 1000;
  console.log(`⏱️ Sequential execution took: ${sequentialDuration.toFixed(2)} seconds\n`);


  // --- SCENARIO 2: CONCURRENT EXECUTION (Parallel Style) ---
  console.log("⚡ [2] Running Tasks CONCURRENTLY (Parallel):");
  const startParallelTime = Date.now();

  // Instead of waiting, we trigger BOTH API calls instantly in the background.
  // promise1 and promise2 start running at the exact same time!
  const promise1 = fetchFromAPI("User Profile (Parallel)", 1500);
  const promise2 = fetchFromAPI("User Orders (Parallel)", 1500);

  // We use `Promise.all` to await both tasks to finish.
  // Since they run at the same time, the total expected time is equal to the SLOWEST task: ~1.5 seconds!
  const [res1, res2] = await Promise.all([promise1, promise2]);
  console.log("  - Successfully fetched (Parallel):", res1.data);
  console.log("  - Successfully fetched (Parallel):", res2.data);

  const parallelDuration = (Date.now() - startParallelTime) / 1000;
  console.log(`⚡ Parallel execution took: ${parallelDuration.toFixed(2)} seconds`);
  console.log("==================================================\n");
}

main();
